### PR TITLE
Zhiqiang/circuit diagram

### PR DESCRIPTION
**Pull request by Xu369**
### Purpose
Output the circuit diagram data(base64 string)

### Demonstrate

- The 'image preview' node could display the diagram

- The 'file' node could save the diagram to an image file on the user's computer

image

### New Requirements

- Modify Node appearance 

- This Node is only for inputs equals to 1, so it needs to be further modified to satisfy any inputs. Make this Node more user-friendly

### CODE REVIEW

**Comment by louislefevre** [comment]
Is this basically just the same code I wrote [in this gist](https://gist.github.com/louislefevre/3949502b5bda0e52ff760b3409bb6c46)? Also the changes to the imports breaks other nodes.

**Comment by louislefevre** [comment]
@Theo-Reignier Do you think we need to add a check so that the node waits for all qubits here?

**Comment by Theo-Reignier** [comment]
> @Theo-Reignier Do you think we need to add a check so that the node waits for all qubits here?

Yes, you are right.
To keep it user-friendly, I think all output nodes should work the same way, meaning that they will be placed the same way in the quantum circuit.
Therefore, this node should include the same 'waiting phase' code as the 'simulator' node

**Comment by louislefevre** [comment]
> I think we need a wilcard import of `qiskit.visualization` as well, since the bloch sphere visualisation requires the methods under this package to work preperly.

We can add that in the PR for the bloch sphere - no need to add it prematurely.

**Comment by Xu369** [comment]
> Is this basically just the same code I wrote [in this gist](https://gist.github.com/louislefevre/3949502b5bda0e52ff760b3409bb6c46)? Also the changes to the imports break other nodes, so I pushed a commit to fix this ([5beef0f](https://github.com/louislefevre/node-red-contrib-quantum/commit/5beef0fa5da7c5678c93bc7a077b2f740d2cef29)). I've also updated some other things and removed some redundancies - take a look at the commit messages for more details.

Yes, because the current method can meet the basic needs, so I create this PR firstly, then seek you guys' opinions, and then I will modify the code and add some other functionalities based on it

File: quantum/nodes/circuit-diagram/circuit-diagram.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+const shell = require('../../python').PythonShell;
+
+module.exports = function(RED) {
+  function CircuitDiagramNode(config) {
+    RED.nodes.createNode(this, config);
+    this.name = config.name;
+
+    this.on('input', async function(msg, send, done) {
+      let script = snippets.CIRCUIT_BUFFER;
+
```
Comment (Theo-Reignier) [inline quantum/nodes/circuit-diagram/circuit-diagram.js:13]: This file should ensure all qubits have finished running before executing the Qiskit script (see 'simulator' node)

Right now, if the file has 2 inputs, he will generate twice the circuit diagram

Comment (Xu369) [reply quantum/nodes/circuit-diagram/circuit-diagram.js:13]: Thank you for your comment, I will modify it

**Review by Theo-Reignier — APPROVED** [review]
`matplotlib` & `plylatenc` should be added to the `venv`

**Comment by louislefevre** [comment]
I'm getting the following error when trying to generate a diagram from the simulator node. Is this normal?
error1

Also, if I connect it to a measure node, there is no output or error message:
error2

**Comment by XxWirusonxX** [comment]
Yes, because there should be no classical bits for the diagram in short explanation.

**Comment by XxWirusonxX** [comment]
I had this problem for my bloch sphere, its probably because of cbit shouldnt exist when performing visualizations.

**Comment by Theo-Reignier** [comment]
> I'm getting the following error when trying to generate a diagram from the simulator node. Is this normal?
> error1
> 
> Also, if I connect it to a measure node, there is no output or error message:
> error2

Louis, these quantum circuits are incorrect, the circuit diagram node should replace the simulator node and wait for all qubits

**Comment by louislefevre** [comment]
> Louis, these quantum circuits are incorrect, the circuit diagram node should replace the simulator node and wait for all qubits

Ah my bad, I thought we were able to connect it to a simulator node. Though I do think the second example should emit some sort of error, if possible.

**Comment by Theo-Reignier** [comment]
> > Louis, these quantum circuits are incorrect, the circuit diagram node should replace the simulator node and wait for all qubits
> 
> Ah my bad, I thought we were able to connect it to a simulator node. Though I do think the second example should emit some sort of error, if possible.

I agree, both situations should throw errors but none of them is supposed to generate an image.

**Comment by Xu369** [comment]
> > Louis, these quantum circuits are incorrect, the circuit diagram node should replace the simulator node and wait for all qubits
> 
> Ah my bad, I thought we were able to connect it to a simulator node. Though I do think the second example should emit some sort of error, if possible.

Yes, it should emit some error message for the second example to remind users that not all Qubits are connected, I'll fix it.
By the way, what about writing these rules in this node's documentation, so that users could know in advance how to use this    @Theo-Reignier

**Comment by Theo-Reignier** [comment]
> > > Louis, these quantum circuits are incorrect, the circuit diagram node should replace the simulator node and wait for all qubits
> > 
> > 
> > Ah my bad, I thought we were able to connect it to a simulator node. Though I do think the second example should emit some sort of error, if possible.
> 
> Yes, it should emit some error message for the second example to remind users that not all Qubits are connected, I'll fix it.
> By the way, what about writing these rules in this node's documentation, so that users could know in advance how to use this @Theo-Reignier

Yes, output nodes will have documentation explaining how to use them. 

By the way, you just said you will add a code to remind the user that not all nodes are connected. How do you plan on implementing this? 
This feature has never been implemented but it would be great for all multiple qubit nodes.

**Comment by Xu369** [comment]
> > > > Louis, these quantum circuits are incorrect, the circuit diagram node should replace the simulator node and wait for all qubits
> > > 
> > > 
> > > Ah my bad, I thought we were able to connect it to a simulator node. Though I do think the second example should emit some sort of error, if possible.
> > 
> > 
> > Yes, it should emit some error message for the second example to remind users that not all Qubits are connected, I'll fix it.
> > By the way, what about writing these rules in this node's documentation, so that users could know in advance how to use this @Theo-Reignier
> 
> Yes, output nodes will have documentation explaining how to use them.
> 
> By the way, you just said you will add a code to remind the user that not all nodes are connected. How do you plan on implementing this?
> This feature has never been implemented but it would be great for all multiple qubit nodes.

I'm thinking about how to do it, and I'll talk to you about it tomorrow if when I have any ideas @Theo-Reignier

**Review by KhesimReid — APPROVED** [review]

**Pull request merged**